### PR TITLE
ci: optimize ci pipeline

### PR DESCRIPTION
**Pull request by imxw**
# Summary

## Description

- Merge some ci pipelines into one because of the order of precedence. For example, if `golang-lint` is failed, there is no need to continue with the remaining steps
- Add go build cache for MacOS (`~/Library/Caches/go-build`)
- Add file filtering so that only changes to these files will trigger the pipeline
- Add conditional judgment for `e2e-test`, only triggered if the repository is `devstream-io/devstream`

### Related Issues

N/A

### Screenshots

image

### Other Information
- The first time will be a bit slow because there is no cache
- Builds on MacOS occasionally have issues that can slow down the overall progress. If this is often the case later, it can be moved to another pipeline

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+            ${{ matrix.go-build }}
+            ~/go/pkg/mod
+          key: ${{ runner.os }}-go-${{ matrix.go }}-${{ hashFiles('**/go.sum') }}
+          restore-keys: |
+            ${{ runner.os }}-go-${{ matrix.go }}-
+            ${{ runner.os }}-go-
+      - name: Test
+        run: go test -v ./...
+      - name: Build
+        run: make build -j8
```
Comment (daniel-hutao) [inline .github/workflows/main.yml:78]: missing new-line

File: .github/workflows/e2e-test.yml [hunk 2]
```diff
@@ -3,28 +3,44 @@ on:
   push:
     branches:
       - main
+    paths:
+      - '**.go'
+      - 'Makefile'
+      - 'go.md'
```
Comment (daniel-hutao) [inline .github/workflows/e2e-test.yml:9]: go.mod?

Comment (imxw) [reply .github/workflows/e2e-test.yml:9]: yes

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -0,0 +1,78 @@
+name: main builder
```
Comment (daniel-hutao) [inline .github/workflows/main.yml:1]: If this file's name changed to `main.yml`, you need to update the reference at README.md(line 8) too. Also notice the README_zh.md

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+on:
+  pull_request:
+    branches: [ main ]
+    paths:
+      - '**.go'
+      - 'Makefile'
+      - 'go.md'
```
Comment (daniel-hutao) [inline .github/workflows/main.yml]: go.mod?

File: .github/workflows/e2e-test.yml [hunk 5]
```diff
@@ -3,28 +3,44 @@ on:
...
-            ~/.cache/go-build
             ~/go/pkg/mod
-          key: ${{ runner.os }}-go-${{ hashFiles('**/go.sum') }}
+            ~/.cache/go-build
+          key: ${{ runner.os }}-go-${{ matrix.go }}-${{ hashFiles('**/go.sum') }}
           restore-keys: |
+            ${{ runner.os }}-go-${{ matrix.go }}-
             ${{ runner.os }}-go-
       - name: Build
         run: make build -j8
```
Comment (daniel-hutao) [inline .github/workflows/e2e-test.yml:45]: If we remove the `push to main` workflow, the ut will never run at main branch. Should we add the ut step here?

Comment (imxw) [reply .github/workflows/e2e-test.yml:45]: The e2e test is triggered when merging into the main branch

**Comment by imxw** [comment]
I found that `lint-commit-message` still has to be split out, because `main.yml` sets up a file filter so that only go-related changes will trigger the pipeline

File: .github/workflows/e2e-test.yml (hunk 5 above)
Comment (imxw) [reply .github/workflows/e2e-test.yml:45]: > If we remove the `push to main` workflow, the ut will never run at main branch. Should we add the ut step here?

`pull request` will trigger ut, how about having `main.yml` support push events?

Comment (daniel-hutao) [reply .github/workflows/e2e-test.yml:45]: Of course, after adding the `push events trigger`, this logic is complete. It's just that I see the simplification you have done. I guess it is to reduce the workflow on the main branch updated, so it seems easier to add `ut` to e2e.yml; if we add `push` directly to main.yml, it seems that `main` and `e2e` cannot avoid repeated `make build`.

File: .github/workflows/lint-commit-message.yml [hunk 6]
```diff
@@ -1,13 +1,13 @@
...
     runs-on: ubuntu-latest
     steps:
-      - uses: actions/checkout@v2
+      - uses: actions/checkout@v3
         with:
           fetch-depth: 0
       - uses: wagoid/commitlint-github-action@v4
         with:
-          configFile: ${{ github.workspace }}/build/ci/commitlint.config.js
+          configFile: ${{ github.workspace }}/build/ci/commitlint.config.js
```
Comment (IronCore864) [inline .github/workflows/lint-commit-message.yml:13]: There is a missing "end of file" at the end of the line. Maybe use an IDE plugin to solve this issue once and for all.

Comment (imxw) [reply .github/workflows/lint-commit-message.yml:13]: Got it, I will fix this issue

**Review by IronCore864 — APPROVED** [review]

**Comment by imxw** [comment]
@daniel-hutao

**Review by daniel-hutao — APPROVED** [review]
/LGTM

**Pull request merged**